File: divide-array-into-arrays-with-max-difference.ts
function divideArray(nums: number[], k: number): number[][] {
    const result: number[][] = [];
    
    nums.sort((a,b) => a - b);

    for (let i = 0; i < nums.length; i += 3) {
        const sliced = nums.slice(i, i  + 3);

        if (sliced[2] - sliced[0] > k)
            return [];

        result.push(sliced);
    }

    return result;
};

console.log(divideArray([1,3,4,8,7,9,3,5,1], 2));
console.log(divideArray([2,4,2,2,5,2], 2));

File: divide-array-into-arrays-with-max-difference.rs
pub fn divide_array(mut nums: Vec<i32>, k: i32) -> Vec<Vec<i32>> {
    let mut r: Vec<Vec<i32>> = Vec::new();

    nums.sort();

    if nums.len() % 3 != 0 {
        return vec![];
    }

    for i in (0..nums.len()).step_by(3) {
        let group = &nums[i..i + 3];

        if group[2] - group[0] > k {
            return vec![];
        }

        r.push(group.to_vec());
    }

    r
}
